### DISCUSSION TITLE
How can I independently adjust channel levels using libvips?

### DISCUSSION THREAD

**Discussion by rafaelmagu** [libvips/libvips#2447] [Q&A] [answered]
Disclaimer: I'm new to image processing altogether, and my Ruby skills are rudimentary at best.

I have a personal project at the moment where I have a list of filters that can be applied to uploaded images to ActiveStorage. Each filter is defined as a helper method. I took inspiration from [this list of filters](https://gist.github.com/tonycoco/2910540), and have modified them to suit this gem. However, I'd like to move away from MiniMagick and towards libvips, and I've been struggling to convert some of the helpers.

For example, I currently have this method using MiniMagick:

```ruby
def filter_lomo(img)
  ImageProcessing::MiniMagick
    .source(img)
    .resize_to_fit(1024, 1024)
    .convert('jpeg')
    .channel('R').level('22%')
    .channel('G').level('22%')
    .saver(strip: true)
    .call
end
```

How can I replicate the above helper using `ImageProcessing::Vips`? Specifically, the `.channel` and `.level calls`, given libvips doesn't have `-channel` nor `-level`. Is there a useful guide for novices like me to better understand this?

I originally asked this question in https://github.com/janko/image_processing/issues/94, where I was then pointed towards this community.

**Comment by jcupitt** [comment, marked as answer]
Hello again, 

libvips images are just big arrays of numbers and it has a full set of operator overloads, so you just write the equation for what you want to do. The intro section in the ruby-vips docs tries to run through the rules:

https://www.rubydoc.info/gems/ruby-vips/Vips

The first example includes what you need --- you can use arrays in expressions to mean pixels, so you can write:

```ruby
image *= [1.22, 1.22, 1]
```

ie. multiply the first and second bands (R and G) by 1.22, and the third (B) by 1.

You can use bandsplit and bandjoin to break bands apart and reassemble them. You can also use ranges in subscripts, ie. `image[0..1]`.

The ruby-vips examples directory has a few code samples:

https://github.com/libvips/ruby-vips/tree/master/example

There's an old  issue on the libvips tracker with some more examples and discussion:

https://github.com/libvips/libvips/issues/59

We've hired an intern to write up a libvips cookbook, so maybe the docs will improve soon.

**Reply by rafaelmagu** [reply]
Thank you, that's excellent.
